### PR TITLE
Add FixZero for LogMeanVariance normalizer

### LINKED ISSUE #2798: Normalize double min and max value returns NaN
### System information

- **OS version/distro**: .Net 4.6.2, Win 10
- **.NET Version (eg., dotnet --info)**: ML.Net nuget 0.10.1

### Issue

- **What did you do?** Input data has a double columns with (double.min, double.max, <100 random numbers from 0 to 10000>), I call 
```
            var normalizeColumns = numericalFeatures.Select(
                f => new NormalizingEstimator.MeanVarColumn(f.Name, fixZero: false, useCdf: false)).ToArray();
            var normalizingEstimator = this.context.Transforms.Normalize(normalizeColumns);
```
I see a feature in transformedData.Preview, which is all NaN, for each row. I use SDCA trainer

- **What happened?** 

The `pipeline.Fit(transformedData)` fails and throw an exception say "train with 0 instances"

- **What did you expect?**

1. ML.Net should handle double.min, double.max for NormalizingEstimator?
2. ML.Net should throw a more meaningful exception - "train with 0 instances" for a feature with NaN is a bit misleading - I do have 100+ rows for this feature. 

### Source code / logs

Please paste or attach the code or logs or traces that would be helpful to diagnose the issue you are reporting.

### PR DESCRIPTION
**Pull request by artidoro**
Fixes #2798.

This PR introduces the `FixZero` argument to the `LogMeanVariance` normalizer, a relative tests and a sample.

It's still in WIP because I would like to make a breaking change instead of creating an overload with required parameter `FixZero`. As soon as the change is accepted and I set up the API Compat tool to accept the breaking change, I should be able to remove the overloads to the `MLContext` extensions.

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVariance.cs [hunk 1]
```diff
@@ -12,15 +12,15 @@ public class NormalizeLogMeanVariance
...
+            // Create a new ML context, for ML.NET operations. It can be used for exception tracking and logging,
             // as well as the source of randomness.
             var mlContext = new MLContext();
             var samples = new List<DataPoint>()
             {
-                new DataPoint(){ Features = new float[4] { 1, 1, 3, 0} },
-                new DataPoint(){ Features = new float[4] { 2, 2, 2, 0} },
-                new DataPoint(){ Features = new float[4] { 0, 0, 1, 0} },
-                new DataPoint(){ Features = new float[4] {-1,-1,-1, 1} }
+                new DataPoint(){ Features = new float[5] { 1, 1, 3, 0, float.MaxValue } },
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVariance.cs]: Is this change related to `FixZero`?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                new DataPoint(){ Features = new float[5] { 0, 0, 1, 0, 0} },
+                new DataPoint(){ Features = new float[5] {-1,-1,-1, 1, 1} }
+            };
+            // Convert training data to IDataView, the general data type used in ML.NET.
+            var data = mlContext.Data.LoadFromEnumerable(samples);
+            // NormalizeLogMeanVariance normalizes the data based on the computed mean and variance of the logarithm of the data.
+            // Uses Cumulative distribution function as output.
+            var normalize = mlContext.Transforms.NormalizeLogMeanVariance(true, "Features", useCdf: true);
+
+            // NormalizeLogMeanVariance normalizes the data based on the computed mean and variance of the logarithm of the data.
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: `NormalizeLogMeanVariance normalizes the data based on the computed mean and variance of the logarithm of the data.` doesn't tell much more than the line below. We need some equations here.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+using System.Linq;
+using Microsoft.ML;
+using Microsoft.ML.Data;
+using static Microsoft.ML.Transforms.NormalizingTransformer;
+
+namespace Samples.Dynamic
+{
+    public class NormalizeLogMeanVarianceFixZero
+    {
+        public static void Example()
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: You have two independent cases in `Example(...)`. Could you split them into two files?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            //  2.0403, 2.0403, 4.0001, 0.0000, 5423991000000000000000000000000000000.0000
+            //  4.0806, 4.0806, 2.6667, 0.0000,-5423991000000000000000000000000000000.0000
+            //  0.0000, 0.0000, 1.3334, 0.0000, 0.0000
+            // -2.0403,-2.0403,-1.3334, 0.0000, 0.0159
+
+            // Let's get transformation parameters. Since we work with only one column we need to pass 0 as parameter for GetNormalizerModelParameters.
+            // If we have multiple columns transformations we need to pass index of InputOutputColumnPair.
+            var transformParams = normalizeTransform.GetNormalizerModelParameters(0) as CdfNormalizerModelParameters<ImmutableArray<float>>;
+            Console.WriteLine("The 1-index value in resulting array would be produce by:");
+            Console.WriteLine($"y = 0.5* (1 + ERF((Math.Log(x)- {transformParams.Mean[1]}) / ({transformParams.StandardDeviation[1]} * sqrt(2)))");
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: We have equation for `columnFixZero` now! May we have the same thing for `column`?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+            // ERF is https://en.wikipedia.org/wiki/Error_function.
+            // Expected output:
+            // The 1 - index value in resulting array would be produce by:
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: What does `1 - index` mean?

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs [hunk 6]
```diff
@@ -159,7 +159,53 @@ public static class NormalizationCatalog
...
+        /// <param name="catalog">The transform catalog</param>
+        /// <param name="fixZero">TODO</param>
+        /// <param name="outputColumnName">Name of the column resulting from the transformation of <paramref name="inputColumnName"/>.
+        ///                                The data type on this column is the same as the input column.</param>
+        /// <param name="inputColumnName">Name of the column to transform. If set to <see langword="null"/>, the value of the <paramref name="outputColumnName"/> will be used as source.
+        ///                               The data type on this column should be <see cref="System.Single"/>, <see cref="System.Double"/> or a known-sized vector of those types.</param>
+        /// <param name="maximumExampleCount">Maximum number of examples used to train the normalizer.</param>
+        /// <param name="useCdf">Whether to use CDF as the output.</param>
+        /// <example>
+        /// <format type="text/markdown">
```
Comment (wschin) [inline src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: Uses would love equations in documentations.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVariance.cs (hunk 1 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVariance.cs]: It's related to the original issue posted by Lisa. They were encountering an issue with the MeanVariance normalizer which does not handle data that looks like `float.MaxValue`, `float.MinValue`. I think it would be useful to show that LogMeanVariance normalizer accepts this data in the related sample.

---
In reply to: [298254220](https://github.com/dotnet/machinelearning/pull/3916#discussion_r298254220) [](ancestors = 298254220)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs (hunk 4 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: I just copied the equation from the LogMeanVariance normalizer sample :) 
There is an equation for both settings of FixZero and for both Cdf and no Cdf. So I think it's complete. 

---
In reply to: [298258190](https://github.com/dotnet/machinelearning/pull/3916#discussion_r298258190) [](ancestors = 298258190)

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs (hunk 6 above)
Comment (artidoro) [reply src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: I will add the equations here as well.

---
In reply to: [298260450](https://github.com/dotnet/machinelearning/pull/3916#discussion_r298260450) [](ancestors = 298260450)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs (hunk 3 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: I have followed the same pattern as the `MeanVariance` normalizer and `LogMeanVariance` normalizer sample. Would you prefer that I split them all into two files? Or should I just split this one? Or keep it like the others for consistency?

---
In reply to: [298257288](https://github.com/dotnet/machinelearning/pull/3916#discussion_r298257288) [](ancestors = 298257288)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs (hunk 2 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: Yes that would be nice I can add that here. 

---
In reply to: [298254677](https://github.com/dotnet/machinelearning/pull/3916#discussion_r298254677) [](ancestors = 298254677)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs (hunk 5 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: I can rephrase here but it just means that we show the normalization parameter values for the slot in the array with index 1.

---
In reply to: [298258499](https://github.com/dotnet/machinelearning/pull/3916#discussion_r298258499) [](ancestors = 298258499)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs (hunk 2 above)
Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: As I told you the other day I am still concerned about that equation. I am going to do a separate PR where I add the equations. (Same for the documentation comment in the catalog that you had).
There is no documentation on how to derive them so I am going through the math again and trying to understand the rational.

---
In reply to: [298311132](https://github.com/dotnet/machinelearning/pull/3916#discussion_r298311132) [](ancestors = 298311132,298254677)

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs (hunk 6 above)
Comment (artidoro) [reply src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: I am a little concerned about the equations. Yesterday I think I saw an issue. I will add those in a separate PR and check this in for now.

---
In reply to: [298310313](https://github.com/dotnet/machinelearning/pull/3916#discussion_r298310313) [](ancestors = 298310313,298260450)

**Comment by artidoro** [comment]
>     /// * Binning - Bucketizes the data in each row and performs a linear rescale based on the calculated bins.

Add equation for log mean variance
both cdf and non cdf

---
Refers to: src/Microsoft.ML.Data/Transforms/Normalizer.cs:46 in 39b4002. [](commit_id = 39b400222de2a20bc4fb01fc5252695ecf7c2402, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs (hunk 3 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeLogMeanVarianceFixZero.cs]: Separating them is more reader-friendly.
